### ISSUE TITLE
Pause Prediction Spikes - a plead for help

### ISSUE THREAD

**Issue by adsorce** [kyutai-labs/delayed-streams-modeling#152]
### Due diligence

- [x] I have done my due diligence in trying to find the answer myself.

### Topic

The Rust implementation

### Question

Firstly, I'd like to say how fun it has been to follow Kyutai's journey with their project releases. As far as I'm concerned, they deserve as much acclaim and fanfare as Sesame got with their Maya release. Excited to see what's still in store for the future.

I have been wrestling with this obsessively for weeks now and I just can't find a solution. It's the last bit of functionality thats missing from making this the perfect platform for my hobby projects, and I would be eternally grateful for some guidance or help. I would happily compensate anyone for their time or raise money for any training needs.

### The Issue
Whenever single digit numbers are spoken, STT has a high probability to try and transcribe them as numerals. This results in the model interpreting the audio as silence which spikes the pause prediction past the 0.6 threshold where it remains even if spoken audio continues. 0.5 second delay after passing the threshold, user turn ends and the TTS responds, cutting off the speaker mid sentence. It's basically impossible to converse about things like phone numbers, addresses, statistics, math etc. The following example shows how the pause prediction remains flat during the beginning of a sentence, then spikes and remains elevated even while audio is continuing. In this case, the streaming is not even showing the individual numeral transcriptions in the word deltas. Sometimes the transcription comes through AFTER the TTS has already begun responding which results in an interruption to it's output as well. Most of the time the transcription never happens at all and occasionally, the whole pipeline will get locked up where no audio can be transcribed until a reset of the WebSocket.

```
PAUSE | Time:  16.62s | Raw: 0.0520 | EMA: 0.0530 | Speaking
PAUSE | Time:  16.70s | Raw: 0.0065 | EMA: 0.0066 | Speaking
WORD  | Time:  16.70s | Text: 'My' | Pause: 0.0066
PAUSE | Time:  16.94s | Raw: 0.0176 | EMA: 0.0175 | Speaking
PAUSE | Time:  17.02s | Raw: 0.0036 | EMA: 0.0037 | Speaking
WORD  | Time:  17.18s | Text: 'number' | Pause: 0.0044
PAUSE | Time:  17.34s | Raw: 0.0319 | EMA: 0.0318 | Speaking
PAUSE | Time:  17.42s | Raw: 0.1502 | EMA: 0.1497 | Speaking
PAUSE | Time:  17.50s | Raw: 0.2051 | EMA: 0.2049 | Speaking
PAUSE | Time:  17.58s | Raw: 0.1311 | EMA: 0.1314 | Speaking
WORD  | Time:  17.58s | Text: 'is' | Pause: 0.1314
PAUSE | Time:  17.66s | Raw: 0.0293 | EMA: 0.0297 | Speaking
PAUSE | Time:  17.90s | Raw: 0.1343 | EMA: 0.1339 | Speaking
PAUSE | Time:  17.98s | Raw: 0.3467 | EMA: 0.3459 | Speaking
PAUSE | Time:  18.06s | Raw: 0.5864 | EMA: 0.5855 | Speaking
PAUSE | Time:  18.14s | Raw: 0.7782 | EMA: 0.7774 | END DETECTED
PAUSE | Time:  18.22s | Raw: 0.9056 | EMA: 0.9051 | END DETECTED
PAUSE | Time:  18.30s | Raw: 0.9219 | EMA: 0.9218 | END DETECTED
PAUSE | Time:  18.38s | Raw: 0.9594 | EMA: 0.9593 | END DETECTED
PAUSE | Time:  18.46s | Raw: 0.9795 | EMA: 0.9794 | END DETECTED
PAUSE | Time:  18.62s | Raw: 0.9909 | EMA: 0.9908 | END DETECTED
PAUSE | Time:  21.10s | Raw: 0.8753 | EMA: 0.8758 | END DETECTED
PAUSE | Time:  21.18s | Raw: 0.8427 | EMA: 0.8429 | END DETECTED
PAUSE | Time:  21.34s | Raw: 0.8880 | EMA: 0.8878 | END DETECTED
PAUSE | Time:  21.42s | Raw: 0.9171 | EMA: 0.9170 | END DETECTED
PAUSE | Time:  21.50s | Raw: 0.8581 | EMA: 0.8583 | END DETECTED
PAUSE | Time:  21.58s | Raw: 0.8391 | EMA: 0.8392 | END DETECTED
PAUSE | Time:  21.66s | Raw: 0.7995 | EMA: 0.7996 | END DETECTED
PAUSE | Time:  21.74s | Raw: 0.9623 | EMA: 0.9617 | END DETECTED
WORD  | Time:  21.74s | Text: '123441.' | Pause: 0.9617
```

Interestingly, if the STT transcribes the numbers as words instead, the pause prediction does not spike and the pipeline flows normally:

```
WORD  | Time:  29.02s | Text: 'My' | Pause: 0.0025
PAUSE | Time:  29.26s | Raw: 0.0241 | EMA: 0.0240 | Speaking
PAUSE | Time:  29.34s | Raw: 0.0105 | EMA: 0.0106 | Speaking
WORD  | Time:  29.58s | Text: 'number' | Pause: 0.0039
PAUSE | Time:  29.74s | Raw: 0.0236 | EMA: 0.0235 | Speaking
PAUSE | Time:  29.82s | Raw: 0.1059 | EMA: 0.1056 | Speaking
PAUSE | Time:  29.90s | Raw: 0.3597 | EMA: 0.3587 | Speaking
PAUSE | Time:  29.98s | Raw: 0.2952 | EMA: 0.2954 | Speaking
PAUSE | Time:  30.06s | Raw: 0.2139 | EMA: 0.2143 | Speaking
WORD  | Time:  30.06s | Text: 'is' | Pause: 0.2143
PAUSE | Time:  30.14s | Raw: 0.1069 | EMA: 0.1073 | Speaking
PAUSE | Time:  30.22s | Raw: 0.0385 | EMA: 0.0388 | Speaking
PAUSE | Time:  30.38s | Raw: 0.0870 | EMA: 0.0868 | Speaking
PAUSE | Time:  30.46s | Raw: 0.1515 | EMA: 0.1513 | Speaking
PAUSE | Time:  30.62s | Raw: 0.0415 | EMA: 0.0419 | Speaking
PAUSE | Time:  30.70s | Raw: 0.0287 | EMA: 0.0288 | Speaking
WORD  | Time:  30.70s | Text: 'one,' | Pause: 0.0288
PAUSE | Time:  30.86s | Raw: 0.0112 | EMA: 0.0113 | Speaking
PAUSE | Time:  31.10s | Raw: 0.0260 | EMA: 0.0260 | Speaking
PAUSE | Time:  31.26s | Raw: 0.0573 | EMA: 0.0571 | Speaking
WORD  | Time:  31.34s | Text: 'two,' | Pause: 0.0650
PAUSE | Time:  31.50s | Raw: 0.0419 | EMA: 0.0420 | Speaking
PAUSE | Time:  31.66s | Raw: 0.0156 | EMA: 0.0157 | Speaking
PAUSE | Time:  31.90s | Raw: 0.0350 | EMA: 0.0349 | Speaking
WORD  | Time:  31.98s | Text: 'three,' | Pause: 0.0279
PAUSE | Time:  32.06s | Raw: 0.0956 | EMA: 0.0953 | Speaking
PAUSE | Time:  32.14s | Raw: 0.0122 | EMA: 0.0125 | Speaking
PAUSE | Time:  32.54s | Raw: 0.0272 | EMA: 0.0272 | Speaking
WORD  | Time:  32.62s | Text: 'four,' | Pause: 0.0253
PAUSE | Time:  32.70s | Raw: 0.0166 | EMA: 0.0166 | Speaking
PAUSE | Time:  32.86s | Raw: 0.0647 | EMA: 0.0645 | Speaking
PAUSE | Time:  32.94s | Raw: 0.3510 | EMA: 0.3499 | Speaking
PAUSE | Time:  33.02s | Raw: 0.5191 | EMA: 0.5185 | Speaking
PAUSE | Time:  33.26s | Raw: 0.6231 | EMA: 0.6227 | END DETECTED
WORD  | Time:  33.26s | Text: 'five.' | Pause: 0.6227
PAUSE | Time:  33.34s | Raw: 0.6803 | EMA: 0.6801 | END DETECTED
PAUSE | Time:  33.42s | Raw: 0.7656 | EMA: 0.7653 | END DETECTED
```

### What I have tried

**Increasing prediction threshold**
Digit audio spikes as high as 0.9999 so there is no value high enough to set it to without completely breaking VAD.

**Increase second pause prediction**
```
https://github.com/kyutai-labs/unmute/blob/d85fbf05a90d2f019a10bf521c1c029d3208b280/unmute/stt/speech_to_text.py#L213
```
Higher values do not mitigate the end user turn

**Relax acceleration of pause prediction**
```
https://github.com/kyutai-labs/unmute/blob/d85fbf05a90d2f019a10bf521c1c029d3208b280/unmute/stt/speech_to_text.py#L88
```
This also has a practically negligible effect

**Set temperature to 0.1 or 0.0**
Hoping to nudge the model into a more deterministic behavior still was unreliable

**Trying to induce word format transcription via prompt injection**
I have tried endless user prompts to possibly prime the model into consistent 'one', 'two', 'three...' style transcription. No luck

**Word delta interception**
I thought that if I could intercept the digit during transcription stream and force a prediction hard reset with each new word it could be a solution. However, being that the deltas for digits don't even exist during streaming, interception is not possible as far as I can tell.

**Remappaing text_logits**
If you inspect the tokenizer mapping for [0:9] you'll find examples like:
`Token sequence [260, 320] → output "1"`
I tried to add a post-processing hook that modifies logits before token sampling to force:
`Token sequence [407] → output "one"`
Implemented on the base model and then converted to candle. No luck either.

**Finetuning begins**
After trying all the low hanging fruit I could think of, I thought for sure we could get a solution via LoRA. I created a 1,000 wav / json dataset targeting  pause prediction behavior with numbers. The first dataset of audio phrases like `I have the series of numbers 4,6,8,etc` or `my zipcode is 10001` had converted transcriptions of all word numbers `i have the series of numbers four, six, eight, etc`. Trained with various configs, all resulted in loss curves showing smooth, gradual convergence, no signs of overfitting.

Still no behavior change.

Then I thought since I structured most of the phrases with numbers at the end of the sentences, it would train the model to continue to treat them as end of user turn. So I remodeled another dataset placing the digits at the beginning or middle of a sentence: `2,3,5,7 are special because they are prime numbers ` etc.

Retrained, still no behavior change.

Then I thought maybe the dataset was still too small to override the pretrained behavior. So I used the entire DailyTalks dataset (after converting any existing digits in the transcriptions to words). Finetuned on that dataset. Nothing. Added my original dataset to the DailyTalks and retrained. Still nothing.

### Conclusion

Apologies for the long write-up but I am incredibly passionate about this. Unfortunately, my coding knowledge and capability seem to not be sufficient at this time. But my willingness to work at this and continue learning is endless. If someone is willing to work on this with me or point me in the right direction, it would go a very long way.

**Comment by vvolhejn** [comment]
I'm sorry to hear that the VAD is not working well for you, clearly you've tried a lot of things.

> I tried to add a post-processing hook that modifies logits before token sampling to force:
> Token sequence [407] → output "one"
> Implemented on the base model and then converted to candle. No luck either.
I'm curious about this one because that's what I would have tried - you implemented the hook in the [Moshi Server codebase](https://github.com/kyutai-labs/moshi/tree/main/rust/moshi-server) as well? Or what do you mean by converted to candle? Just checking that you're not just converting the weights.

As a workaround, you can try to disable the VAD head altogether and replace it with [Silero VAD](https://github.com/snakers4/silero-vad). Silero is not semantic, it just tells you whether a person is talking or not. But if you fiddle with the EMA parameters you can make it predict a pause e.g. 1 second after the person has stopped talking, independent of what's being said. Obviously, that will increase the latency by quite a bit, but perhaps it would be an acceptable trade-off for your usecase?

**Comment by vvolhejn** [comment]
Actually, I think I know why this happens. The VAD is trained to predict whether there will be a word added to the transcript within the next 2 seconds. When you start dictating numbers one by one, there are many cases where you will say something like "three eight seven one", which will take you more than 2 seconds to say and gets transcribed to "3871" once you've said everything.

Therefore, the VAD is matching the training objective well - if I say:
```
My
phone
number
is
3871
```
the delay between "is" and "3121" is >2 seconds, so the VAD correctly predicts end of speech, according to the definition it was trained with.

Currently we haven't had other people reporting issues with the VAD so this is unlikely to get fixed very soon. It would require retraining the VAD with a better training objective. Hopefully this at least gives you an idea of what's going wrong.

**Comment by adsorce** [comment]
> Actually, I think I know why this happens. The VAD is trained to predict whether there will be a word added to the transcript within the next 2 seconds. When you start dictating numbers one by one, there are many cases where you will say something like "three eight seven one", which will take you more than 2 seconds to say and gets transcribed to "3871" once you've said everything.
> 
> Therefore, the VAD is matching the training objective well - if I say:
> ```
> My
> phone
> number
> is
> 3871
> ```
> the delay between "is" and "3121" is >2 seconds, so the VAD correctly predicts end of speech, according to the definition it was trained with.
> 
> Currently we haven't had other people reporting issues with the VAD so this is unlikely to get fixed very soon. It would require retraining the VAD with a better training objective. Hopefully this at least gives you an idea of what's going wrong.

Yes, this does make sense because for numbers like eleven or ninety nine or even dollar amounts like one thousand, there is a specific expectation of the length of the audio and there is no issue. Random single digit numbers may confuse that expectation because there is no trained estimation. 

Btw, you can test this issue out directly on unmute.sh. I just tried again and out of 8 websocket restarts only 1 resolved towards the model consistently transcribing numbers into words. 

To answer your question about the candle conversion, I was manipulating the weights on the PyTorch model and then converting to candle for the Rust implementation. I realize now that I could have just worked with the weights directly in the candle model.

The initial way I approached remapping was to permanently modify the text_linear.weight matrix. Setting a higher multiplier to the number word tokens to ensure they appeared with a higher probability than number digits during greedy decoding. This didn't work because I later discovered there are sometimes negative probabilities so any kind of multiplier would still be low. Then I hardcoded large negative values for the number digits to try and guarantee they never appeared as a high probability token. This completely broke the model for reasons I don't know.

The second remapping approach was instead of manipulating the weights directly, I tried a runtime remapping during inference itself. I vendored Kyutai’s moshi crate, added a helper in vendor/moshi/src/asr.rs that zeroes out the digit token logits before sampling, and rewired the sampling step to call it. Then I pointed moshi-server at that local crate so cargo install --path vendor/moshi-server produced a moshi binary with the runtime remapping baked in. This also broke the model. I feel like I gave up too early with this approach as I was already exhausted. I'm going to take another crack at this approach. 

On another note, function/tool calling in the unmute pipeline is such a pleasure to implement and works better than any other application I've tried.

**Comment by vvolhejn** [comment]
> The initial way I approached remapping was to permanently modify the text_linear.weight matrix. Setting a higher multiplier to the number word tokens to ensure they appeared with a higher probability than number digits during greedy decoding. This didn't work because I later discovered there are sometimes negative probabilities so any kind of multiplier would still be low. Then I hardcoded large negative values for the number digits to try and guarantee they never appeared as a high probability token. This completely broke the model for reasons I don't know.

"Negative probabilities" - they're logits, so they're pre-softmax probabilities and maybe positive or negative. So multiplying a row (column?) by a constant might increase or decrease the probability depending on the sign. See [code](https://github.com/0x53504852/moshi-rs/blob/44a08bbb60bb289e77e3ac219acd05f560e9dbeb/moshi-core/src/lm.rs#L899).

> that zeroes out the digit token logits before sampling, and rewired the sampling step to call it.

If you want to prevent them from being sampled, you want to -inf those logits rather than zero them, since that will lead to zero probability after the softmax.

> On another note, function/tool calling in the unmute pipeline is such a pleasure to implement and works better than any other application I've tried.

Thank you! What and how did you implement? We'd be happy to accept a PR in Unmute that shows a proof of concept function calling, because this is definitely a very common requirement and we don't have it in Unmute yet, see https://github.com/kyutai-labs/unmute/issues/77

**Comment by adsorce** [comment]
When I get a chance, I can clean up the function calling implementation and post it in unmute.

I made some progress on the remapping using neg_infinity and leaving the word logits untouched. The model is now always transcribing in the way I want "one two three...". However end-turn behavior was still interrupting mid sentence. My theory is that I'm only intervening at the sampler. The transformer still emits the same hidden state it would have used to pick the digit tokens, and the pause “extra head” is computed from that hidden state before we touch the logits. So even though the decoded text is “one two three...”, the internal representation still looks like a digit sequence, and the VAD head keeps spiking.

Going to see if its possible to rerun the forward pass after the word token is chosen and use that new transformer output for pause prediction. Will probably introduce a bit of latency but at least it would verify what's happening.

**Comment by vvolhejn** [comment]
Hm, that actually makes sense since the VAD is ahead of the text stream as you say. What's interesting is what you reported before, where the VAD would not spike if the transcription is in words and not digits. I guess the model reaches an internal state where it decides you're not about to dictate some long number and so you're still speaking, but the internal state precedes the text token chosen, so the intervention doesn't help.

**Comment by sjswerdloff** [comment]
I've had it split the words Four Thousand (spoken) in to 
4
00
0

While I understand splitting the four from the thousand, I'm at a loss to understand why "thousand" would get converted with separate "words" of zeros (and not all three zeros).
None-the-less: Kudos to Kyutai, it's incredibly performant and accurate on MLX (I've incorporated a 1 second additional delay to accumulate words that might otherwise get split with VAD and increased pr_vad to 0.7).

**Comment by fcollonval** [comment]
> When I get a chance, I can clean up the function calling implementation and post it in unmute.

@adsorce would you be willing to share your implementation of the tools handling in unmute?
